### PR TITLE
Add a customized io.Copy, wrapped in struct io.Copier

### PR DESCRIPTION
**Pull request by Soulou**
* To use in `tarball` package
* To use in `go-fssync` package

Allowing large IO operations without impacting file system disk cache

Specs are not flaky on my system:

```
└> go test -count=1 ./...
ok  	github.com/Scalingo/go-utils/io	1.753s

└> go test -count=1 ./...
ok  	github.com/Scalingo/go-utils/io	2.077s

└> go test -count=1 ./...
ok  	github.com/Scalingo/go-utils/io	1.942s

└> go test -count=1 ./...
ok  	github.com/Scalingo/go-utils/io	1.913s

└> go test -count=10 ./...
ok  	github.com/Scalingo/go-utils/io	53.295s
```

If I remove the NoCache flag

```
└> go test -count=5 ./... 
--- FAIL: TestCopier_Copy (0.45s)
    --- FAIL: TestCopier_Copy/it_should_not_impact_cache_if_option_is_set (0.45s)
        copier_test.go:101: 
            	Error Trace:	copier_test.go:101
            	Error:      	Should be true
            	Test:       	TestCopier_Copy/it_should_not_impact_cache_if_option_is_set
            	Messages:   	expected diff of cache < 50MB, was 99MB
--- FAIL: TestCopier_Copy (0.41s)
    --- FAIL: TestCopier_Copy/it_should_not_impact_cache_if_option_is_set (0.41s)
        copier_test.go:101: 
            	Error Trace:	copier_test.go:101
            	Error:      	Should be true
            	Test:       	TestCopier_Copy/it_should_not_impact_cache_if_option_is_set
            	Messages:   	expected diff of cache < 50MB, was 100MB
--- FAIL: TestCopier_Copy (0.43s)
    --- FAIL: TestCopier_Copy/it_should_not_impact_cache_if_option_is_set (0.43s)
        copier_test.go:101: 
            	Error Trace:	copier_test.go:101
            	Error:      	Should be true
            	Test:       	TestCopier_Copy/it_should_not_impact_cache_if_option_is_set
            	Messages:   	expected diff of cache < 50MB, was 100MB
--- FAIL: TestCopier_Copy (0.43s)
    --- FAIL: TestCopier_Copy/it_should_not_impact_cache_if_option_is_set (0.43s)
        copier_test.go:101: 
            	Error Trace:	copier_test.go:101
            	Error:      	Should be true
            	Test:       	TestCopier_Copy/it_should_not_impact_cache_if_option_is_set
            	Messages:   	expected diff of cache < 50MB, was 100MB
--- FAIL: TestCopier_Copy (0.42s)
    --- FAIL: TestCopier_Copy/it_should_not_impact_cache_if_option_is_set (0.42s)
        copier_test.go:101: 
            	Error Trace:	copier_test.go:101
            	Error:      	Should be true
            	Test:       	TestCopier_Copy/it_should_not_impact_cache_if_option_is_set
            	Messages:   	expected diff of cache < 50MB, was 100MB
FAIL
FAIL	github.com/Scalingo/go-utils/io	2.144s
FAIL
```

it looks pretty good to me

### CODE REVIEW

**Comment by Soulou** [comment]
Arf afraid it doesn't work in a container..

**Comment by Soulou** [comment]
So in codeship. What do you think @johnsudaar ?

**Comment by johnsudaar** [comment]
Do you know why ? We will run it from a container (in dev at least maybe one day in prod). It would be nice to know why it fails on those environments.

**Comment by Soulou** [comment]
I don't know codeship setup, I jsut tried in a ctonaienr locally:

```
└> docker run -v `pwd`:/mnt -it golang:1.14 bash     
...
root@56a2523994fd:/mnt# go test -count=20  ./...
ok  	github.com/Scalingo/go-utils/io	51.335s
```
It works as expected..

**Comment by Soulou** [comment]
@johnsudaar using Travis instead of codeship? :D

File: io/copier.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	for {
+		nr, er := src.Read(buf)
+		if nr > 0 {
+			if fdSrc, ok := src.(Fder); c.noDiskCacheRead && ok {
+				// Fadvise is a system call giving instruction to the OS about how to behave
+				// with the flag FADC_DONTNEED, it tell the OS to drop the disk cache
+				// on a given file, on a give part of the file (initial offset + end offset)
+				// http://man7.org/linux/man-pages/man2/posix_fadvise.2.html
+				unix.Fadvise(int(fdSrc.Fd()), written, written+int64(nr), unix.FADV_DONTNEED)
+			}
```
Comment (johnsudaar) [inline io/copier.go:68]: I don't get it what's the point to run Fadvise on the part already read ? Shouldn't this be done before the read ?

File: io/copier.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+			nw, ew := dst.Write(buf[0:nr])
+			if nw > 0 {
+				if fdDst, ok := dst.(Fder); c.noDiskCacheWrite && ok {
+					unix.Fadvise(int(fdDst.Fd()), written, written+int64(nw), unix.FADV_DONTNEED)
+				}
```
Comment (johnsudaar) [inline io/copier.go:74]: Same here ?

File: io/copier.go (hunk 1 above)
Comment (Soulou) [reply io/copier.go:68]: Nope, fadvise is designed to free the cache memory actually, so you first read a chunk, then discard it, and again and again

File: io/copier.go (hunk 2 above)
Comment (Soulou) [reply io/copier.go:74]: Same explanation

**Pull request merged**